File: Leetcode/typescript/sol377.ts

//https://leetcode.com/problems/combination-sum-iv/description/

function combinationSum4(nums: number[], target: number): number {
    const dp:number[] = Array(target+1).fill(0);
    dp[0] = 1;
    for(let i = 1; i < dp.length; i++){
        for(const num of nums){
            if(i - num >= 0){
                dp[i] += dp[i-num]
            }
        }
    }
    return dp.at(-1);
};

File: Leetcode/python/sol377.py
# https://leetcode.com/problems/combination-sum-iv/
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        memo = {}
        def helper(val):
            if val in memo:
                return memo[val]
            if val == 0:
                return 1
            if val < 0:
                return 0
            count = 0
            for i in nums:
                count += helper(val - i)
            memo[val] = count
            return count
        return helper(target)
